pkg/controller/cloneset/update/cloneset_update.go

```diff
@@ -185,51 +257,72 @@ func (c *realControl) updatePod(cs *appsv1alpha1.CloneSet, coreControl clonesetc
+					clonesetutils.UpdateExpectations.ExpectUpdated(clonesetutils.GetControllerKey(cs), updateRevision.Name, pod)
+					return res.DelayDuration, nil
+				}
+
+				c.recorder.Eventf(cs, v1.EventTypeWarning, "FailedUpdatePodInPlace", "failed to update pod %s in-place(revision %v): %v", pod.Name, updateRevision.Name, res.UpdateErr)
+				return res.DelayDuration, res.UpdateErr
+			}
 		}
 
 		if cs.Spec.UpdateStrategy.Type == appsv1alpha1.InPlaceOnlyCloneSetUpdateStrategyType {
-			return res.DelayDuration, fmt.Errorf("find Pod %s update strategy is InPlaceOnly but can not update in-place", pod.Name)
+			return 0, fmt.Errorf("find Pod %s update strategy is InPlaceOnly but can not update in-place", pod.Name)
 		}
 		klog.Warningf("CloneSet %s/%s can not update Pod %s in-place, so it will back off to ReCreate", cs.Namespace, cs.Name, pod.Name)
 	}
 
-	klog.V(2).Infof("CloneSet %s/%s deleting Pod %s for update %s", cs.Namespace, cs.Name, pod.Name, updateRevision.Name)
+	klog.V(2).Infof("CloneSet %s/%s start to patch Pod %s specified-delete for update %s", cs.Namespace, cs.Name, pod.Name, updateRevision.Name)
 
-	c.scaleExp.ExpectScale(clonesetutils.GetControllerKey(cs), expectations.Delete, pod.Name)
-	if err := c.Delete(context.TODO(), pod); err != nil {
-		c.scaleExp.ObserveScale(clonesetutils.GetControllerKey(cs), expectations.Delete, pod.Name)
+	if patched, err := specifieddelete.PatchPodSpecifiedDelete(c, pod, "true"); err != nil {
 		c.recorder.Eventf(cs, v1.EventTypeWarning, "FailedUpdatePodReCreate",
-			"failed to delete pod %s for update: %v", pod.Name, err)
+			"failed to patch pod specified-delete %s for update(revision %s): %v", pod.Name, updateRevision.Name, err)
 		return 0, err
+	} else if patched {
+		clonesetutils.ResourceVersionExpectations.Expect(pod)
 	}
 
-	// TODO(FillZpp): add a strategy controlling if the PVCs of this pod should be deleted
-	for _, pvc := range pvcs {
-		if pvc.Labels[appsv1alpha1.CloneSetInstanceID] != pod.Labels[appsv1alpha1.CloneSetInstanceID] {
-			continue
-		}
-
-		c.scaleExp.ExpectScale(clonesetutils.GetControllerKey(cs), expectations.Delete, pvc.Name)
-		if err := c.Delete(context.TODO(), pvc); err != nil {
-			c.scaleExp.ObserveScale(clonesetutils.GetControllerKey(cs), expectations.Delete, pvc.Name)
-			c.recorder.Eventf(cs, v1.EventTypeWarning, "FailedDelete", "failed to delete pvc %s: %v", pvc.Name, err)
-			return 0, err
-		}
-	}
+	//clonesetutils.ScaleExpectations.ExpectScale(clonesetutils.GetControllerKey(cs), expectations.Delete, pod.Name)
+	//if err := c.Delete(context.TODO(), pod); err != nil {
+	//	clonesetutils.ScaleExpectations.ObserveScale(clonesetutils.GetControllerKey(cs), expectations.Delete, pod.Name)
+	//	c.recorder.Eventf(cs, v1.EventTypeWarning, "FailedUpdatePodReCreate",
+	//		"failed to delete pod %s for update: %v", pod.Name, err)
+	//	return 0, err
+	//}
+	//
+	//// TODO(FillZpp): add a strategy controlling if the PVCs of this pod should be deleted
+	//for _, pvc := range pvcs {
```

Comment:
Is there any reason the pvc deletion code is commented?

Reply:
We patch `apps.kruise.io/specified-delete` label to pod here, and the scale module will then delete this Pod and its PVCs.